snapcraft/internal/meta/_snap_packaging.py

```diff
@@ -120,17 +120,34 @@ def _update_yaml_with_extracted_metadata(
         if not part:
             raise meta_errors.AdoptedPartMissingError(part_name)
 
-        # This would be caught since metadata would be missing, but we want
-        # to be clear about the issue here. This really should be caught by the
-        # schema, but it doesn't seem to support such dynamic behavior.
-        if 'parse-info' not in config_data['parts'][part_name]:
-            raise meta_errors.AdoptedPartNotParsingInfo(part_name)
-
-        # Get the metadata from the pull step first, then update it using the
-        # metadata from the build step (i.e. the data from the build step takes
-        # precedence over the pull step)
-        metadata = part.get_pull_state().extracted_metadata['metadata']
-        metadata.update(part.get_build_state().extracted_metadata['metadata'])
+        pull_state = part.get_pull_state()
```

Comment:
this should probably be improved upon in the future with a `part.get_state()` call

Reply:
I'm not sure I understand, what would `part.get_state()` do?

Reply:
Bring in a `state` object for the part for which we could do `state.pull_state.extracted_metadata['metadata']`

Reply:
Or maybe encapsulate the knowledge of that dict too